File: src/main/java/de/seine_eloquenz/open_glass/OpenGlassServer.java
package de.seine_eloquenz.open_glass;


import de.seine_eloquenz.open_glass.endpoints.Endpoint;
import de.seine_eloquenz.open_glass.endpoints.EndpointNotFound;
import de.seine_eloquenz.open_glass.endpoints.EndpointPressKey;
import de.seine_eloquenz.open_glass.game.Game;
import de.seine_eloquenz.open_glass.game.Games;
import fi.iki.elonen.NanoHTTPD;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Logger;

public class OpenGlassServer extends NanoHTTPD {

    /**
     * logger to log to.
     */
    public static final Logger LOG = Logger.getLogger(OpenGlassServer.class.getName());
    private static final Response FORBIDDEN
            = newFixedLengthResponse(Response.Status.FORBIDDEN, MIME_PLAINTEXT, "You're not authenticated");

    private final String apiKey;
    private final Map<String, Endpoint> endpoints;
    private Game game;
    private final GameProvider gameProvider;

    public OpenGlassServer(final String hostname, final int port, final String apiKey) {
        super(hostname, port);
        this.endpoints = new HashMap<>();
        this.apiKey = apiKey;
        this.game = Games.NONE;
        gameProvider = () -> game;
        this.endpoints.put("/presskey", new EndpointPressKey(gameProvider));
    }

    @Override
    public Response serve(final IHTTPSession session) {
        Method method = session.getMethod();
        String uri = session.getUri();
        OpenGlassServer.LOG.info(method + " '" + uri + "' ");
        Map<String, List<String>> params = session.getParameters();
        if (params.get("API_KEY") == null || !apiKey.equals(params.get("API_KEY").get(0))) {
            return FORBIDDEN;
        } else {
            if ("/chooseGame".equals(uri)) {
                if (params.containsKey("game")) {
                    try {
                        this.game = Games.valueOf(params.get("game").get(0));
                        return Endpoint.OK;
                    } catch (IllegalArgumentException e) {
                        return Endpoint.BAD_REQUEST;
                    }
                } else {
                    return Endpoint.BAD_REQUEST;
                }
            }
            return endpoints.getOrDefault(uri, new EndpointNotFound()).serve(params);
        }
    }
}

File: open-glass-server/src/main/java/de/seine_eloquenz/open_glass/server/OpenGlassServer.kt
package de.seine_eloquenz.open_glass.server

import de.seine_eloquenz.open_glass.server.endpoints.Endpoint
import de.seine_eloquenz.open_glass.server.endpoints.EndpointKey.Factory.endpointKeyHold
import de.seine_eloquenz.open_glass.server.endpoints.EndpointKey.Factory.endpointKeyPress
import de.seine_eloquenz.open_glass.server.endpoints.EndpointNotFound
import de.seine_eloquenz.open_glass.server.endpoints.EndpointStatus
import de.seine_eloquenz.open_glass.server.game.Game
import de.seine_eloquenz.open_glass.server.game.Games
import fi.iki.elonen.NanoHTTPD
import java.util.*
import java.util.logging.Logger

/**
 * Main Server class handling HTTP requests
 */
class OpenGlassServer(hostname: String?, port: Int, apiKey: String) : NanoHTTPD(hostname, port) {

    companion object {

        /**
         * logger to log to.
         */
        @JvmField
        val LOG: Logger = Logger.getLogger(OpenGlassServer::class.java.name)
    }

    private val apiKey: String
    private val endpoints: MutableMap<String, Endpoint>
    private var game: Game
    private val gameProvider: GameProvider
    private var stop: Boolean = false

    init {
        endpoints = HashMap()
        this.apiKey = apiKey
        game = Games.NONE
        gameProvider = GameProvider { game }
        endpoints["/presskey"] = endpointKeyPress(gameProvider)
        endpoints["/holdkey"] = endpointKeyHold(gameProvider, 3000)
        endpoints["/status"] = EndpointStatus(gameProvider)
    }

    override fun serve(session: IHTTPSession): Response {
        val method = session.method
        val uri = session.uri
        LOG.info("$method '$uri' ")
        val params = session.parameters
        return if (params["API_KEY"] == null || apiKey != params["API_KEY"]!![0]) {
            Endpoint.FORBIDDEN
        } else {
            return when (uri) {
                "/chooseGame" -> {
                    if (params.containsKey("game")) {
                        try {
                            game = Games.valueOf(params["game"]!![0])
                            Endpoint.OK
                        } catch (e: IllegalArgumentException) {
                            Endpoint.BAD_REQUEST
                        }
                    } else {
                        Endpoint.BAD_REQUEST
                    }
                }
                "/shutdown" -> {
                    stop = true
                    LOG.info("Received shutdown signal. Will shutdown on next Main tick.")
                    Endpoint.OK
                }
                else -> {
                    endpoints.getOrDefault(uri, EndpointNotFound()).serve(params)
                }
            }

        }
    }

    /**
     * Returns whether the server received a shutdown signal and is about to stop
     */
    fun shallStop(): Boolean = stop
}